src/System.Threading.Timer/tests/TimerFiringTests.cs

```diff
@@ -0,0 +1,110 @@
+        {
+            are.Set();
+        }), null, TimeSpan.FromMilliseconds(250), TimeSpan.FromMilliseconds(Timeout.Infinite) /* not relevant */))
+        {
+            Assert.True(are.WaitOne(TimeSpan.FromMilliseconds(750)));
+        }
+    }
+
+    [Fact]
+    public void Timer_Fires_AndPassesStateThroughCallback()
+    {
+        using (var t = new Timer(new TimerCallback((object s) =>
+        {
+            Assert.NotNull(s);
+        }), new object(), TimeSpan.FromMilliseconds(100), TimeSpan.FromMilliseconds(Timeout.Infinite) /* not relevant */)) { }
+    }
+
+    [Fact]
+    public void Timer_Fires_AndPassesNullStateThroughCallback()
+    {
+        using (var t = new Timer(new TimerCallback((object s) =>
+        {
+            Assert.Null(s);
+        }), null, TimeSpan.FromMilliseconds(100), TimeSpan.FromMilliseconds(Timeout.Infinite) /* not relevant */)) { }
+    }
+
+    [Fact]
+    public void Timer_Fires_After_DueTime_AndOn_Period()
+    {
+        int count = 0;
+        AutoResetEvent are = new AutoResetEvent(false);
+
+        using (var t = new Timer(new TimerCallback((object s) =>
+        {
+            count++;
+            if (count >= 2)
+            {
+                are.Set();
+            }
+        }), null, TimeSpan.FromMilliseconds(500), TimeSpan.FromMilliseconds(50)))
+        {
+            Assert.True(are.WaitOne(TimeSpan.FromMilliseconds(2000)));
+        }
+    }
+
+    [Fact]
+    public void Timer_FiresOnlyOnce_OnDueTime_With_InfinitePeriod()
+    {
+        int count = 0;
+        AutoResetEvent are = new AutoResetEvent(false);
+
+        using (var t = new Timer(new TimerCallback((object s) =>
+        {
+            count++;
+            if (count >= 2)
+            {
+                are.Set();
+            }
+        }), null, TimeSpan.FromMilliseconds(100), TimeSpan.FromMilliseconds(Timeout.Infinite) /* not relevant */))
+        {
+            Assert.False(are.WaitOne(TimeSpan.FromSeconds(1)));
+        }
+    }
+
+    [Fact]
+    public void MultipleTimers_ShouldFire_InCorrectOrder()
+    {
+        Object t1State = new Object();
+        Object t2State = new Object();
+        Timer t1 = null;
+        Timer t2 = null;
+        int count = 0;
+        AutoResetEvent are = new AutoResetEvent(false);
+        TimerCallback tc = new TimerCallback((object o) =>
+        {
+            if (count == 0)
+            {
+                Assert.Same(t2State, o);
+                count++;
+            }
+            else
+            {
+                Assert.Same(t1State, o);
+                are.Set();
+            }
+        });
+
+        using (t1 = new Timer(tc, t1State, TimeSpan.FromSeconds(1), TimeSpan.FromMilliseconds(-1)))
+        using (t2 = new Timer(tc, t2State, TimeSpan.FromMilliseconds(0), TimeSpan.FromMilliseconds(-1)))
+        {
+            // Wait for both events to fire
+            Assert.True(are.WaitOne(TimeSpan.FromMilliseconds(1500 /*1.5 seconds*/)));
+        }
+    }
+}
```

Comment:
Some other tests to consider:
- An outer loop test that verifies the Timer can be garbage collected when no one's referencing it, and that it stops firing in that case
- An outer loop test that verifies a repeated timer still fires/executes subsequent invocations even if the first one is blocked.
- A test that verifies the Timer can dispose itself from its own callback
- A test that verifies the Timer can be disposed multiple times
- A test that verifies we can create lots (thousands) of timers and have many timers firing at the same time, e.g. create ten thousand timers that signal a countdown event and that have a one millisecond timeout
- A test that verifies a non-repeating timer that's already fired can have its timeout changed to be restarted